src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/HashProviderDispenser.Unix.cs

```diff
@@ -13,38 +13,47 @@ internal static partial class HashProviderDispenser
     {
         public static HashProvider CreateHashProvider(string hashAlgorithmId)
         {
-            switch (hashAlgorithmId)
-            {
-                case HashAlgorithmNames.SHA1:
-                    return new EvpHashProvider(Interop.Crypto.EvpSha1());
-                case HashAlgorithmNames.SHA256:
-                    return new EvpHashProvider(Interop.Crypto.EvpSha256());
-                case HashAlgorithmNames.SHA384:
-                    return new EvpHashProvider(Interop.Crypto.EvpSha384());
-                case HashAlgorithmNames.SHA512:
-                    return new EvpHashProvider(Interop.Crypto.EvpSha512());
-                case HashAlgorithmNames.MD5:
-                    return new EvpHashProvider(Interop.Crypto.EvpMd5());
-            }
-            throw new CryptographicException(SR.Format(SR.Cryptography_UnknownHashAlgorithm, hashAlgorithmId));
+            IntPtr evpType = HashAlgorithmToEvp(hashAlgorithmId);
+            return new EvpHashProvider(evpType);
         }
 
-        public static unsafe HashProvider CreateMacProvider(string hashAlgorithmId, ReadOnlySpan<byte> key)
+        public static HashProvider CreateMacProvider(string hashAlgorithmId, ReadOnlySpan<byte> key)
         {
-            switch (hashAlgorithmId)
+            IntPtr evpType = HashAlgorithmToEvp(hashAlgorithmId);
+            return new HmacHashProvider(evpType, key);
+        }
+
+        private static IntPtr HashAlgorithmToEvp(string hashAlgorithmId) => hashAlgorithmId switch {
+            HashAlgorithmNames.SHA1 => Interop.Crypto.EvpSha1(),
+            HashAlgorithmNames.SHA256 => Interop.Crypto.EvpSha256(),
+            HashAlgorithmNames.SHA384 => Interop.Crypto.EvpSha384(),
+            HashAlgorithmNames.SHA512 => Interop.Crypto.EvpSha512(),
+            HashAlgorithmNames.MD5 => Interop.Crypto.EvpMd5(),
+            _ => throw new CryptographicException(SR.Format(SR.Cryptography_UnknownHashAlgorithm, hashAlgorithmId))
```

Comment:
Since these are shared handles we can probably gain some mild wins by memoizing the values in managed (and avoiding the P/Invoke every time to get a redundant answer).